File: tools/fleet-dashboard/server/guideQueue.js
'use strict';

// Guide Queue — same file-kanban model as tasks.js (mirrored on purpose),
// for the guide-writing pipeline: ops/guide-queue/{ideas,drafted,ready,
// released,rejected}/*.md. Backs the Fleet Dashboard's Guides tab.
//
// See tools/guide-queue/lib/guide_queue.py — the Python side the cron roles
// use. This is the same frontmatter contract read/written from Node instead.

const fs = require('node:fs');
const path = require('node:path');
const yaml = require('js-yaml');
const { siteDir } = require('./sites');

const STATUSES = ['ideas', 'drafted', 'ready', 'released', 'rejected'];

const QUEUE_FIELDS = [
  'queue_id', 'queue_status', 'source', 'created', 'brief', 'notes',
  'hero_image', 'card_image',
];

const FIELD_ORDER = [...QUEUE_FIELDS, 'title', 'description', 'category',
  'updated', 'published', 'author', 'featuredProducts', 'faq', 'schemaType', 'steps'];

const SCHEMA = yaml.CORE_SCHEMA; // same rationale as tasks.js: keep `created: 2026-08-10` a string

function isValidStatus(s) { return STATUSES.includes(s); }

// Same filename-safety rule as tasks.js.
function isValidFilename(name) {
  return typeof name === 'string' && /^[A-Za-z0-9][A-Za-z0-9._-]*\.md$/.test(name) && !name.includes('..');
}

function queueDir(root, slug) { return path.join(siteDir(root, slug), 'ops', 'guide-queue'); }
function statusDir(root, slug, status) { return path.join(queueDir(root, slug), status); }

function parseItem(text) {
  const m = text.match(/^---\r?\n([\s\S]*?)\r?\n---\r?\n?([\s\S]*)$/);
  if (!m) return { meta: {}, body: text };
  let meta = {};
  try { meta = yaml.load(m[1], { schema: SCHEMA }) || {}; } catch { meta = {}; }
  if (typeof meta !== 'object' || Array.isArray(meta)) meta = {};
  return { meta, body: m[2] };
}

function serializeItem(meta, body) {
  const clean = {};
  for (const k of Object.keys(meta || {})) {
    const v = meta[k];
    if (v === undefined || v === null || v === '') continue;
    if (Array.isArray(v) && v.length === 0) continue;
    clean[k] = v;
  }
  const ordered = {};
  for (const k of FIELD_ORDER) if (k in clean) { ordered[k] = clean[k]; delete clean[k]; }
  for (const k of Object.keys(clean)) ordered[k] = clean[k];
  const cleanBody = (body || '').replace(/^\n+/, '').replace(/\n+$/, '');
  const fm = yaml.dump(ordered, { schema: SCHEMA, lineWidth: 100, quotingType: '"', forceQuotes: false }).trimEnd();
  return `---\n${fm}\n---\n\n${cleanBody}\n`;
}

function readItemCard(dir, status, name, slug) {
  if (!name.endsWith('.md')) return null;
  const fp = path.join(dir, name);
  let st; try { st = fs.statSync(fp); } catch { return null; }
  if (!st.isFile()) return null;
  const { meta, body } = parseItem(fs.readFileSync(fp, 'utf8'));
  const excerpt = body.replace(/^#.*$/gm, '').replace(/\s+/g, ' ').trim().slice(0, 200);
  return {
    site: slug, file: name, status,
    title: meta.title || meta.brief && String(meta.brief).slice(0, 60) || name.replace(/\.md$/, ''),
    category: meta.category || null,
    source: meta.source || null,
    created: meta.created ? String(meta.created) : null,
    hasImages: !!(meta.hero_image && meta.card_image),
    excerpt,
    mtime: st.mtimeMs,
  };
}

// List every item across all statuses for one site.
function list(root, slug) {
  const base = queueDir(root, slug);
  const out = {};
  for (const status of STATUSES) {
    out[status] = [];
    const dir = path.join(base, status);
    let names = [];
    try { names = fs.readdirSync(dir); } catch { names = []; }
    for (const name of names.sort()) {
      const card = readItemCard(dir, status, name, slug);
      if (card) out[status].push(card);
    }
  }
  return out;
}

// Aggregate across every site with a guide-queue dir.
function listAll(root, slugs) {
  const all = [];
  for (const slug of slugs) {
    const base = queueDir(root, slug);
    if (!fs.existsSync(base)) continue;
    for (const status of STATUSES) {
      const dir = path.join(base, status);
      let names = [];
      try { names = fs.readdirSync(dir); } catch { continue; }
      for (const name of names.sort()) {
        const card = readItemCard(dir, status, name, slug);
        if (card) all.push(card);
      }
    }
  }
  return all;
}

// Full item content incl. body + resolvable image paths, for the preview modal.
function get(root, slug, status, file) {
  if (!isValidStatus(status) || !isValidFilename(file)) throw httpErr(400, 'bad status or filename');
  const fp = path.join(statusDir(root, slug, status), file);
  if (!fs.existsSync(fp)) throw httpErr(404, 'item not found');
  const raw = fs.readFileSync(fp, 'utf8');
  const { meta, body } = parseItem(raw);
  const images = {};
  for (const [key, field] of [['hero', 'hero_image'], ['card', 'card_image']]) {
    if (meta[field]) {
      const abs = path.join(siteDir(root, slug), meta[field]);
      images[key] = fs.existsSync(abs)
        ? `/api/guide-queue/${slug}/image?path=${encodeURIComponent(meta[field])}`
        : null;
    }
  }
  return { file, status, meta, body, images };
}

// Serve raw image bytes for the preview modal. `relPath` must resolve inside
// this site's ops/guide-queue/drafted-assets/ or site/public/images/ — the
// two places generate-guide-images.py / guide-publish.sh ever write to.
function imagePath(root, slug, relPath) {
  if (typeof relPath !== 'string' || relPath.includes('..')) throw httpErr(400, 'bad path');
  const base = siteDir(root, slug);
  const abs = path.resolve(base, relPath);
  const allowedRoots = [
    path.resolve(base, 'ops', 'guide-queue', 'drafted-assets'),
    path.resolve(base, 'site', 'public', 'images'),
  ];
  if (!allowedRoots.some((r) => abs === r || abs.startsWith(r + path.sep))) {
    throw httpErr(400, 'path outside allowed image roots');
  }
  if (!fs.existsSync(abs)) throw httpErr(404, 'image not found');
  return abs;
}

function slugify(s) {
  return String(s || 'guide').toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-+|-+$/g, '').slice(0, 60) || 'guide';
}

function today() {
  const d = new Date();
  const p = (n) => String(n).padStart(2, '0');
  return `${d.getFullYear()}-${p(d.getMonth() + 1)}-${p(d.getDate())}`;
}

// Add a new idea (dashboard's "add idea" form — same shape as the AI
// seeder's add-idea CLI call, just from Node instead of Python).
function addIdea(root, slug, payload) {
  const title = payload.title || 'Untitled guide idea';
  const qid = slugify(title);
  const meta = {
    queue_id: qid,
    queue_status: 'idea',
    source: payload.source === 'ai' ? 'ai' : 'human',
    created: today(),
    brief: payload.brief || '',
    title,
    category: payload.category || undefined,
  };
  const dir = statusDir(root, slug, 'ideas');
  fs.mkdirSync(dir, { recursive: true });
  let name = `${qid}.md`, n = 2;
  while (fs.existsSync(path.join(dir, name))) { name = `${qid}-${n}.md`; n += 1; }
  fs.writeFileSync(path.join(dir, name), serializeItem(meta, ''));
  return name;
}

// Move between columns — used by the dashboard's Accept/Reject/Send-back
// buttons. Plain fs rename; the next cron tick (or a human) picks up state
// from the directory, same non-committing convention as tasks.move().
function move(root, slug, fromStatus, file, toStatus) {
  if (!isValidStatus(fromStatus) || !isValidStatus(toStatus) || !isValidFilename(file)) {
    throw httpErr(400, 'bad status or filename');
  }
  const from = path.join(statusDir(root, slug, fromStatus), file);
  if (!fs.existsSync(from)) throw httpErr(404, 'item not found');
  const toDir = statusDir(root, slug, toStatus);
  fs.mkdirSync(toDir, { recursive: true });
  let dest = path.join(toDir, file), name = file, n = 2;
  while (fs.existsSync(dest)) { name = file.replace(/\.md$/, `-${n}.md`); dest = path.join(toDir, name); n += 1; }
  // Keep queue_status in sync with its new directory.
  const raw = fs.readFileSync(from, 'utf8');
  const { meta, body } = parseItem(raw);
  meta.queue_status = { ideas: 'idea', drafted: 'drafted', ready: 'ready', released: 'released', rejected: 'rejected' }[toStatus] || toStatus;
  fs.writeFileSync(from, serializeItem(meta, body));
  fs.renameSync(from, dest);
  return { status: toStatus, file: name };
}

// Overwrite notes / a light edit from the preview modal (title/brief/notes
// only — the actual guide body is the writer role's job, not a dashboard
// text field).
function update(root, slug, status, file, payload) {
  if (!isValidStatus(status) || !isValidFilename(file)) throw httpErr(400, 'bad status or filename');
  const fp = path.join(statusDir(root, slug, status), file);
  if (!fs.existsSync(fp)) throw httpErr(404, 'item not found');
  const { meta, body } = parseItem(fs.readFileSync(fp, 'utf8'));
  if (payload.notes !== undefined) meta.notes = payload.notes;
  if (payload.title !== undefined) meta.title = payload.title;
  if (payload.brief !== undefined) meta.brief = payload.brief;
  if (payload.category !== undefined) meta.category = payload.category;
  fs.writeFileSync(fp, serializeItem(meta, body));
  return { file };
}

// --- Per-site config (ops/tracked.yaml's manual: block) -------------------
// Read via full YAML parse (safe); WRITE via a targeted line replace so we
// never clobber the hand-written comments in tracked.yaml (a full
// yaml.dump round-trip would strip them — this file is meant to stay
// human-readable, same reasoning as tasks.js's comment about CORE_SCHEMA).

function trackedYamlPath(root, slug) { return path.join(siteDir(root, slug), 'ops', 'tracked.yaml'); }

function getConfig(root, slug) {
  const fp = trackedYamlPath(root, slug);
  let data = {};
  try { data = yaml.load(fs.readFileSync(fp, 'utf8'), { schema: SCHEMA }) || {}; } catch { data = {}; }
  const manual = (data && data.manual) || {};
  return {
    guide_cadence_days: manual.guide_cadence_days != null ? Number(manual.guide_cadence_days) : 5,
    guide_ideas_min: manual.guide_ideas_min != null ? Number(manual.guide_ideas_min) : 3,
  };
}

function setConfigField(root, slug, field, value) {
  if (!['guide_cadence_days', 'guide_ideas_min'].includes(field)) throw httpErr(400, 'unknown config field');
  const n = Number(value);
  if (!Number.isFinite(n) || n < 1) throw httpErr(400, 'value must be a positive number');
  const fp = trackedYamlPath(root, slug);
  if (!fs.existsSync(fp)) throw httpErr(404, 'ops/tracked.yaml not found for this site');
  let text = fs.readFileSync(fp, 'utf8');
  // Replace only the value token, preserving any trailing `# comment` on the
  // same line (tracked.yaml's fields are hand-documented — a full-line
  // replace would silently drop that documentation on every dashboard edit).
  const lineRe = new RegExp(`^(\\s*${field}:)\\s*\\S+(\\s*(#.*)?)$`, 'm');
  if (lineRe.test(text)) {
    text = text.replace(lineRe, `$1 ${n}$2`);
  } else if (/^manual:\s*$/m.test(text)) {
    text = text.replace(/^manual:\s*$/m, `manual:\n  ${field}: ${n}`);
  } else {
    text += `\nmanual:\n  ${field}: ${n}\n`;
  }
  fs.writeFileSync(fp, text);
  return getConfig(root, slug);
}

function httpErr(status, msg) { const e = new Error(msg); e.httpStatus = status; return e; }

module.exports = {
  STATUSES, list, listAll, get, imagePath, addIdea, move, update,
  getConfig, setConfigField,
  parseItem, serializeItem,
};

File: tools/guide-queue/lib/guide_queue.py
"""tools/guide-queue/lib/guide_queue.py — shared file-kanban queue for the
fleet's guide-writing pipeline: idea -> drafted -> ready -> released -> rejected.

One directory tree per site, at `ops/guide-queue/<status>/<id>.md`. Mirrors
tools/fleet-dashboard/server/tasks.js's frontmatter-markdown shape on purpose
(YAML frontmatter + markdown body, flat filenames, per-status subdirs) so the
dashboard's existing task-board code and mental model extend cleanly to guides
instead of inventing a second convention.

A `drafted`/`ready` item's body IS the future guide's markdown body — the
frontmatter carries both queue-only fields (queue_id, queue_status, source,
brief, notes, hero_image, card_image) and the real guide frontmatter fields
(title, description, category, ...). guide_publisher.strip_queue_fields()
removes the queue-only keys when copying an item into the site's live content
collection at publish time.

Mounted read-only into site worker containers via the same
`.monorepo-tools` bind that already exposes tools/task-budget etc (see
docker-compose.yml). No third-party deps beyond PyYAML (already required by
several other fleet tools mounted the same way).
"""
from __future__ import annotations

import datetime
import re
from pathlib import Path
from typing import Any

import yaml

STATUSES = ["ideas", "drafted", "ready", "released", "rejected"]

# Queue-only frontmatter keys — never copied into the live site content file.
QUEUE_FIELDS = [
    "queue_id",
    "queue_status",
    "source",
    "created",
    "brief",
    "notes",
    "hero_image",
    "hero_alt",
    "card_image",
]

# Preferred key order when serializing (queue fields first, then guide fields).
_FIELD_ORDER = QUEUE_FIELDS + [
    "title",
    "description",
    "category",
    "updated",
    "published",
    "author",
    "featuredProducts",
    "faq",
    "schemaType",
    "steps",
]

_FM_RE = re.compile(r"^---\r?\n(.*?)\r?\n---\r?\n?(.*)$", re.S)


class GuideQueueError(RuntimeError):
    pass


def queue_dir(site_root: Path) -> Path:
    return Path(site_root) / "ops" / "guide-queue"


def _status_dir(site_root: Path, status: str) -> Path:
    if status not in STATUSES:
        raise GuideQueueError(f"bad status: {status!r}")
    return queue_dir(site_root) / status


def parse_item(text: str) -> tuple[dict, str]:
    m = _FM_RE.match(text)
    if not m:
        return {}, text
    try:
        meta = yaml.safe_load(m.group(1)) or {}
    except yaml.YAMLError as e:
        # 2026-08-31: this used to swallow the error and return {} — a
        # writer role that emits a frontmatter value with an unescaped quote
        # (e.g. `lede: "Search bait" is not...`, a plain scalar starting
        # with a quoted phrase) produced invalid YAML, and every field
        # silently vanished instead of failing loudly. Worse, a downstream
        # metadata-restore step would then treat the empty {} as the item's
        # real current state and discard every already-correct field
        # (tags, relatedCategory, hero_image, ...) it didn't own — silent
        # data loss, not just a validation gap. Raise instead: every caller
        # (cli.py, guide_writer_meta.py) already has a clear error path for
        # this and a broken file staying visibly broken is much safer than
        # one that quietly reads back empty.
        raise GuideQueueError(f"invalid YAML frontmatter: {e}") from e
    if not isinstance(meta, dict):
        raise GuideQueueError("frontmatter did not parse to a mapping")
    return meta, m.group(2)


def serialize_item(meta: dict, body: str) -> str:
    clean = {k: v for k, v in meta.items() if v not in (None, "", [])}
    ordered: dict[str, Any] = {}
    for k in _FIELD_ORDER:
        if k in clean:
            ordered[k] = clean.pop(k)
    ordered.update(clean)  # any unexpected keys survive, appended at the end
    fm = yaml.safe_dump(
        ordered, sort_keys=False, allow_unicode=True, default_flow_style=False, width=100
    ).strip()
    body_clean = (body or "").strip("\n")
    tail = f"\n{body_clean}\n" if body_clean else "\n"
    return f"---\n{fm}\n---\n{tail}"


def slugify(s: str) -> str:
    s = re.sub(r"[^a-z0-9]+", "-", (s or "").lower()).strip("-")
    return s[:60] or "guide"


def _read(fp: Path) -> tuple[dict, str]:
    return parse_item(fp.read_text(encoding="utf-8"))


def list_status(site_root: Path, status: str) -> list[dict]:
    d = _status_dir(site_root, status)
    items = []
    if d.is_dir():
        for f in sorted(d.glob("*.md")):
            # A single corrupt file (unparseable frontmatter) must not sink
            # the whole scan — `oldest()` backs the writer's "is anything
            # pending?" gate, and one broken drafted item would otherwise
            # block every other idea/site from ever being picked up. Surface
            # it as a flagged, meta-less item instead so callers that care
            # (get_item on that exact file, or a human scanning the list)
            # still see it broke, without wedging everything behind it.
            try:
                meta, body = _read(f)
            except GuideQueueError as e:
                items.append(
                    {
                        "file": f.name,
                        "status": status,
                        "meta": {},
                        "excerpt": "",
                        "parse_error": str(e),
                    }
                )
                continue
            items.append(
                {
                    "file": f.name,
                    "status": status,
                    "meta": meta,
                    "excerpt": re.sub(r"\s+", " ", body).strip()[:200],
                }
            )
    return items


def list_all(site_root: Path) -> dict[str, list[dict]]:
    return {s: list_status(site_root, s) for s in STATUSES}


def get_item(site_root: Path, status: str, filename: str) -> dict:
    fp = _status_dir(site_root, status) / filename
    if not fp.is_file():
        raise GuideQueueError(f"not found: {status}/{filename}")
    meta, body = _read(fp)
    return {"file": filename, "status": status, "meta": meta, "body": body}


def add_idea(
    site_root: Path,
    title: str,
    brief: str,
    category: str,
    source: str = "human",
) -> str:
    """Append a new idea to ideas/. Returns the created filename."""
    qid = slugify(title)
    meta = {
        "queue_id": qid,
        "queue_status": "idea",
        "source": source,
        "created": datetime.date.today().isoformat(),
        "brief": brief,
        "title": title,
        "category": category,
    }
    d = _status_dir(site_root, "ideas")
    d.mkdir(parents=True, exist_ok=True)
    name, n = f"{qid}.md", 2
    while (d / name).exists():
        name = f"{qid}-{n}.md"
        n += 1
    (d / name).write_text(serialize_item(meta, ""), encoding="utf-8")
    return name


def write_item(site_root: Path, status: str, filename: str, meta: dict, body: str) -> None:
    """Overwrite an item's frontmatter + body in place (used by the writer role
    after it fills in the drafted guide content)."""
    fp = _status_dir(site_root, status) / filename
    if not fp.is_file():
        raise GuideQueueError(f"not found: {status}/{filename}")
    fp.write_text(serialize_item(meta, body), encoding="utf-8")


def move_item(site_root: Path, from_status: str, filename: str, to_status: str) -> str:
    src = _status_dir(site_root, from_status) / filename
    if not src.is_file():
        raise GuideQueueError(f"not found: {from_status}/{filename}")
    dst_dir = _status_dir(site_root, to_status)
    dst_dir.mkdir(parents=True, exist_ok=True)
    name, n = filename, 2
    dst = dst_dir / name
    while dst.exists():
        name = filename.replace(".md", f"-{n}.md")
        dst = dst_dir / name
        n += 1
    # Update queue_status in the frontmatter to match its new home so the two
    # never drift (directory is the source of truth for the dashboard's list
    # views; the field matters for anything reading a single file in isolation,
    # e.g. the publisher after a move).
    meta, body = _read(src)
    meta["queue_status"] = {
        "ideas": "idea",
        "drafted": "drafted",
        "ready": "ready",
        "released": "released",
        "rejected": "rejected",
    }.get(to_status, to_status)
    src.write_text(serialize_item(meta, body), encoding="utf-8")
    src.rename(dst)
    return name


def oldest(site_root: Path, status: str) -> dict | None:
    """The oldest item in a status column by `created` date (falls back to
    filename order for items missing that field).

    Corrupt frontmatter is surfaced by ``list_status`` for the dashboard, but
    it is not a publishable candidate.  Do not let a flagged item become the
    queue head and make the publisher fail before it can reach valid items
    behind it.
    """
    items = [it for it in list_status(site_root, status) if not it.get("parse_error")]
    items.sort(key=lambda it: (it["meta"].get("created") or "", it["file"]))
    return items[0] if items else None


def strip_queue_fields(meta: dict) -> dict:
    """Drop queue-only keys, leaving pure guide frontmatter for the live
    content collection. `hero_image`/`card_image` are handled separately by
    the publisher (renamed to the collection's `image`/`cardImage` fields
    after the files are copied into site/public/)."""
    return {k: v for k, v in meta.items() if k not in QUEUE_FIELDS}


def existing_titles(site_root: Path) -> set[str]:
    """Every title already in the pipeline (any status) — for the idea-seeder
    to avoid proposing a duplicate."""
    titles = set()
    for status in STATUSES:
        for it in list_status(site_root, status):
            t = it["meta"].get("title")
            if t:
                titles.add(str(t).strip().lower())
    return titles
